### PR TITLE
refactor(key-gen): persist intermediates in DB for key-gen

### PR DESCRIPTION
**Pull request by 0xThemis**
- **build: updated taceo crates**
- **ci: ignore RUSTSEC-2023-0089 from circom-witness-rs**
- **wip**

---

> [!NOTE]
> **Medium Risk**
> Touches key-gen protocol state handling and persistence semantics (new DB schema, transactional finalize/abort/delete behavior), which can affect liveness and correctness if edge cases are missed; changes are well-covered by added/updated tests but still impact core protocol flow.
> 
> **Overview**
> Key-gen/reshare now **persists in-progress protocol state** via the secret manager so runs can resume across restarts. This introduces `KeyGenIntermediateValues` (serializable secret material), a new `in_progress_keygens` Postgres table (plus trigger), and updates `KeyGenPoly`/ephemeral keys to support canonical (de)serialization.
> 
> The key-gen flow is refactored so `DLogSecretGenService` becomes async and uses the secret manager for round-1 intermediates, round-3 pending shares, and finalize confirmation; `key_event_watcher` is updated accordingly (including clearer abort/delete semantics). The `SecretManager` trait is expanded (abort/delete intermediates, stage pending shares, confirm finalized shares with epoch rollback protection), with a substantial Postgres implementation update using transactional/idempotent `confirm_dlog_share` and retry logic; tests are rewritten/added to cover idempotency and failure modes.
> 
> Minor: bumps `rustls-webpki`, removes obsolete on-chain contribution structs from `oprf-types`, and tweaks `release-plz` commit grouping.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 56c5ad2f5163edc0456b7dcd1741d254598b5932. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: oprf-key-gen/src/services/key_event_watcher/tests.rs [hunk 1]
```diff
@@ -13,22 +31,169 @@ use oprf_types::{
...
+            .lock()
+            .expect("lock not poisoned")
+            .retain(|(key_id, _), _| *key_id != oprf_key_id);
+        self.pending_shares
+            .lock()
+            .expect("lock not poisoned")
+            .retain(|(key_id, _), _| *key_id != oprf_key_id);
+        Ok(())
+    }
+}
```
[the remark this answers is not in the document]
Comment (0xThemis) [reply oprf-key-gen/src/services/key_event_watcher/tests.rs:205]: follow up PR removes this:
https://github.com/TaceoLabs/oprf-service/pull/565

File: oprf-core/src/keygen.rs [hunk 2]
```diff
@@ -39,7 +40,7 @@ const COEFF_DS: &[u8] = b"KeyGenPolyCoeff";
 /// During key generation, each party creates shares of this polynomial for their nodes. When all individual shares from different parties are accumulated, the resulting shares correspond to the final shared secret key (the sum of individual secrets).
 ///
 /// Parties shall forget the polynomial after key-generation.
-#[derive(ZeroizeOnDrop)]
+#[derive(Clone, ZeroizeOnDrop, CanonicalSerialize, CanonicalDeserialize)]
```
Comment (dkales) [inline oprf-core/src/keygen.rs]: Do we need this to be Clone?

File: oprf-key-gen/migrations/20260416133133_keygen_intermediates_table.up.sql [hunk 3]
```diff
@@ -0,0 +1,21 @@
+-- Add up migration script here
+
+CREATE TABLE intermediates (
```
Comment (dkales) [inline oprf-key-gen/migrations/20260416133133_keygen_intermediates_table.up.sql]: probably could use a better name?

File: oprf-key-gen/migrations/20260416133133_keygen_intermediates_table.up.sql [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    pending_share BYTEA,
+    intermediates BYTEA NOT NULL,
+
+    created_at TIMESTAMPTZ NOT NULL DEFAULT now(),
+    updated_at TIMESTAMPTZ NOT NULL DEFAULT now(),
+
+    PRIMARY KEY (id, pending_epoch)
+);
+
+CREATE INDEX idx_intermediates_id ON intermediates(id);
```
Comment (dkales) [inline oprf-key-gen/migrations/20260416133133_keygen_intermediates_table.up.sql]: Do we need an index, and if yes, do we want ones on other tables as well?

File: Cargo.toml [hunk 5]
```diff
@@ -53,7 +53,6 @@ metrics = "0.24"
 nodes-common = { package = "taceo-nodes-common", version = "0.4.3", default-features = false }
 nodes-observability = { package = "taceo-nodes-observability", version = "0.1" }
 num-bigint = "0.4"
-parking_lot = "0.12"
 poseidon2 = { package = "taceo-poseidon2", version = "0.2", default-features = false }
```
Comment (dkales) [inline Cargo.toml:57]: Is removing parking_lot worth it?

File: oprf-test-utils/src/secret_manager/test_secret_manager.rs [hunk 6]
```diff
@@ -172,6 +191,54 @@ impl TestSecretManager {
...
+
+    /// Confirms the pending share for the given key ID, making it the active share.
+    ///
+    /// Moves the pending share from `pending_shares` into `store`, paired with
+    /// the provided epoch and public key.
+    ///
+    /// # Panics
+    ///
+    /// Panics if no pending share exists for `oprf_key_id`.
+    pub async fn confirm_dlog_share(
```
Comment (dkales) [inline oprf-test-utils/src/secret_manager/test_secret_manager.rs:230]: I know this will probably not happen since we do not use it this way, but the DB method uses a TX over the whole operation, while this is doing 2 locks, which is not the same behavior. Maybe just have both maps under 1 lock?

File: oprf-key-gen/src/services/secret_manager.rs [hunk 7]
```diff
@@ -12,45 +12,106 @@ use async_trait::async_trait;
...
-    async fn remove_oprf_key_material(&self, oprf_key_id: OprfKeyId) -> eyre::Result<()>;
+    /// During ordinary flow it is not possible to have multiple key-gens running at the same time for the same key. Theoretically though a node might drop out during an ongoing key-gen and come back later, trying to start a new run. If we then abort, we still want to remove all in-progress state associated with this key.
+    async fn abort_keygen(&self, oprf_key_id: OprfKeyId) -> Result<()>;
 
-    /// Stores an OPRF secret with with the provided epoch.
+    /// Tries to persist the intermediate values needed for key generation (or reshare).
     ///
-    /// This method SHOULD overwrite/delete the old epoch, if it already existed.
-    async fn store_dlog_share(
+    /// If intermediate values already exist for this `OprfKeyId` and `ShareEpoch` pair, this method must return the already store `KeyGenIntermediateValues` and discard the new ones.
```
Comment (dkales) [inline oprf-key-gen/src/services/secret_manager.rs]:
```suggestion
    /// If intermediate values already exist for this `OprfKeyId` and `ShareEpoch` pair, this method must return the already stored `KeyGenIntermediateValues` and discard the new ones.
```

Comment (dkales) [reply oprf-key-gen/src/services/secret_manager.rs]: Also, is this for sure what we want?

File: oprf-key-gen/src/services/secret_gen.rs [hunk 8]
```diff
@@ -57,42 +61,34 @@ pub(crate) enum Contributions {
...
     key_gen_material: Arc<CircomGroth16Material>,
 }
 
+/// Intermediate values generated during the key-generation (or reshare) protocol.
+///
+/// These are persisted between protocol rounds via the [`SecretManager`](crate::services::secret_manager::SecretManager) and consumed once the round completes.
+#[derive(Clone, CanonicalSerialize, CanonicalDeserialize)]
+pub struct KeyGenIntermediateValues {
+    sk: EphemeralEncryptionPrivateKey,
+    poly: Option<KeyGenPoly>,
```
Comment (dkales) [inline oprf-key-gen/src/services/secret_gen.rs:81]: This could use a comment on when it is None, hat to think a bit about it.

File: oprf-key-gen/src/services/secret_gen.rs [hunk 9]
```diff
@@ -142,229 +141,215 @@ impl ToxicWasteRound1 {
...
-            .remove(&oprf_key_id)
-            .context("Did not have round 1 toxic waste stored")?;
+    ) -> eyre::Result<Option<SecretGenCiphertexts>> {
+        let intermediate_values = self
+            .secret_manager
+            .fetch_keygen_intermediates(oprf_key_id, pending_epoch)
+            .await
+            .context("while fetching intermediate values")?;
+        let Some(intermediate_values) = intermediate_values else {
+            tracing::trace!("no intermediate values stored for round 2 - this key-gen is stuck");
```
Comment (dkales) [inline oprf-key-gen/src/services/secret_gen.rs]: Do we want to emit an error at least? This would go unnoticed now, right?

File: oprf-key-gen/src/services/secret_gen.rs [hunk 10]
```diff
@@ -142,229 +141,215 @@ impl ToxicWasteRound1 {
...
-            .remove(&oprf_key_id)
-            .context("Did not have round 2 toxic waste stored")?;
-        let share = decrypt_key_gen_ciphertexts(ciphers, toxic_waste_r2, sharing_type, pks)
+        let intermediate_values = self
+            .secret_manager
+            .fetch_keygen_intermediates(oprf_key_id, pending_epoch)
+            .await
+            .context("while fetching intermediate values")?;
+        let Some(intermediate_values) = intermediate_values else {
+            tracing::trace!("no intermediate values stored for round 3 - this key-gen is stuck");
```
Comment (dkales) [inline oprf-key-gen/src/services/secret_gen.rs]: same here?

File: oprf-key-gen/src/services/secret_manager/postgres.rs [hunk 11]
```diff
@@ -249,5 +502,16 @@ fn to_db_ark_serialize_uncompressed<T: CanonicalSerialize>(t: &T) -> Vec<u8> {
     bytes
 }
 
+#[inline]
+#[allow(
+    clippy::needless_pass_by_value,
+    reason = "nicer ergonomics at callsite"
+)]
+fn from_db_ark_serialize_uncompressed<T: CanonicalDeserialize>(b: Vec<u8>) -> Result<T> {
```
Comment (dkales) [inline oprf-key-gen/src/services/secret_manager/postgres.rs:510]: Cannot directly make commentson the above, but the serialization should be wrapped in a zeroize so we at least try as much as possible to honor the zeroize property of the things we are storing here.

Wrap serialized buffers with `zeroize::Zeroizing<Vec<u8>>` on both the encode and decode paths. sqlx can bind `&[u8]`, so `Zeroizing<Vec<u8>>` should slot in with `.as_slice()` or a deref.

File: oprf-key-gen/src/services/secret_manager/postgres.rs [hunk 12]
```diff
@@ -227,19 +346,153 @@ impl PostgresSecretManager {
...
+    async fn with_retry<F, Fut, T>(&self, op_name: &str, f: F) -> Result<T>
+    where
+        F: Fn() -> Fut,
+        Fut: Future<Output = Result<T>>,
+    {
+        f.retry(self.backoff_strategy())
+            .sleep(tokio::time::sleep)
+            .when(is_retryable_error)
+            .notify(|e, duration| {
+                tracing::warn!("Retrying {op_name} in db: {e:?} after {duration:?}");
```
Comment (dkales) [inline oprf-key-gen/src/services/secret_manager/postgres.rs]: Flagged by AI review: maybe we should not print the full error as it might leak bindings?

File: oprf-key-gen/src/services/key_event_watcher.rs [hunk 13]
```diff
@@ -207,12 +203,9 @@ async fn handle_events(args: KeyEventWatcherTaskConfig) -> eyre::Result<()> {
...
     log: Log<LogData>,
     contract: &OprfKeyRegistryInstance<DynProvider>,
-    secret_gen: &mut DLogSecretGenService,
-    secret_manager: &SecretManagerService,
+    secret_gen: &DLogSecretGenService,
     transaction_handler: &TransactionHandler,
 ) -> Result<()> {
-    // First step during every log is asynchronously ping the secret-manager to wake up in case it is in deep-sleep
-    ping_secret_manager(secret_manager.to_owned(), tracing::Span::current());
     let result = match log.topic0() {
```
Comment (dkales) [inline oprf-key-gen/src/services/key_event_watcher.rs:209]: If we no longer ping, will this lead to problems/long latencies for keygen? Might need to investigate the DB setup soon.

File: oprf-key-gen/src/services/secret_manager/postgres.rs [hunk 14]
```diff
@@ -227,19 +346,153 @@ impl PostgresSecretManager {
...
+        .bind(to_db_ark_serialize_uncompressed(share))
+        // Postgres lacks u32; cast to i64 to satisfy SQLx type mapping
+        .bind(i64::from(pending_epoch))
+        .bind(to_db_ark_serialize_uncompressed(public_key))
+        .execute(conn)
+        .await?
+        .rows_affected())
+    }
+
+    async fn delete_intermediates_inner(
```
Comment (dkales) [inline oprf-key-gen/src/services/secret_manager/postgres.rs:446]: This should probably also take the epoch, no?

Claude flagged a scenario where it could happen due to some event replay, idk if we want to defend against this:
```
The tests verify idempotency at the SecretManager API level but none simulate the scenario from Issue 1: a cross-epoch interleaving where a finalize-replay wipes in-progress intermediates for a later epoch. Add a test:

  1. Confirm share for (id, N).
  2. Store intermediates for (id, N+1).
  3. Call confirm_dlog_share(id, N, pk) again (idempotent replay).
  4. Assert intermediates for (id, N+1) are still present.
```

File: oprf-key-gen/src/services/secret_manager/postgres.rs [hunk 15]
```diff
@@ -59,163 +101,240 @@ impl PostgresSecretManager {
...
+
+    #[instrument(level = "debug", skip(self))]
+    async fn delete_oprf_key_material(&self, oprf_key_id: OprfKeyId) -> super::Result<()> {
+        tracing::trace!("trying to delete key-material..");
+
+        let delete_transaction = || async {
+            let mut tx = self.pool.begin().await?;
+            let conn = tx.acquire().await?;
+            // use standard isolation level: READ COMMITTED
+            sqlx::query("SET TRANSACTION ISOLATION LEVEL READ COMMITTED")
```
Comment (dkales) [inline oprf-key-gen/src/services/secret_manager/postgres.rs:146]: this is the default though, do we want to explicitly set it?

File: oprf-key-gen/src/services/secret_manager/postgres.rs [hunk 16]
```diff
@@ -59,163 +101,240 @@ impl PostgresSecretManager {
...
-                        public_key = EXCLUDED.public_key
-                    WHERE
-                        shares.epoch < EXCLUDED.epoch AND
-                        shares.deleted = false;
-                ",
+        let confirm_dlog_share = || async {
+            let mut tx = self.pool.begin().await?;
+            let conn = tx.acquire().await?;
+            // Use isolation level: SERIALIZABLE
+            sqlx::query("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE")
```
Comment (dkales) [inline oprf-key-gen/src/services/secret_manager/postgres.rs:305]: this could use more explanation on why it is used.

File: oprf-key-gen/src/services/secret_manager/postgres.rs [hunk 17]
```diff
@@ -59,163 +101,240 @@ impl PostgresSecretManager {
...
+
+        let delete_transaction = || async {
+            let mut tx = self.pool.begin().await?;
+            let conn = tx.acquire().await?;
+            // use standard isolation level: READ COMMITTED
+            sqlx::query("SET TRANSACTION ISOLATION LEVEL READ COMMITTED")
+                .execute(&mut *conn)
+                .await?;
+            // Soft-delete finalized shares for this key.
+            let deleted_shares = Self::soft_delete_shares_inner(oprf_key_id, &mut *tx).await?;
```
Comment (dkales) [inline oprf-key-gen/src/services/secret_manager/postgres.rs:150]: Are these two deletes not an error as they should never live in the first place? I guess it is fine to delete them anyways though.

File: oprf-key-gen/src/services/secret_manager/postgres.rs [hunk 18]
```diff
@@ -59,163 +101,240 @@ impl PostgresSecretManager {
...
         }
+        Ok(maybe_intermediates)
     }
 
     #[instrument(level = "debug", skip(self))]
-    async fn remove_oprf_key_material(&self, oprf_key_id: OprfKeyId) -> eyre::Result<()> {
-        tracing::trace!("trying to delete key-material..");
-        let rows_deleted = (|| {
-            sqlx::query(
+    async fn abort_keygen(&self, oprf_key_id: OprfKeyId) -> super::Result<()> {
```
Comment (dkales) [inline oprf-key-gen/src/services/secret_manager/postgres.rs:237]: maybe this should also be epoch scoped?

File: oprf-core/src/keygen.rs (hunk 2 above)
Comment (0xThemis) [reply oprf-core/src/keygen.rs]: Only for the tests - I can do a `cfg_attr(test)` good point

File: oprf-key-gen/migrations/20260416133133_keygen_intermediates_table.up.sql (hunk 3 above)
Comment (0xThemis) [reply oprf-key-gen/migrations/20260416133133_keygen_intermediates_table.up.sql]: Maybe `key_gen_staging`? I wanted to move from toxic_waste but maybe still better name... Open for ideas

File: Cargo.toml (hunk 5 above)
Comment (0xThemis) [reply Cargo.toml:57]: Maybe not - can drop the commit is a dedicated one but after reading some blog posts I was in favor of std locking again 😅

File: oprf-key-gen/src/services/secret_gen.rs (hunk 9 above)
Comment (0xThemis) [reply oprf-key-gen/src/services/secret_gen.rs]: Yes I wanted to use the upgraded proxy contract to report stuck key-gens in a follow-up PR therefore I left this as todo basically. All errors from the DB are also silently wrapped into 
 `eyre::Report` currently and not consumed but can do that at least

File: oprf-key-gen/src/services/secret_manager/postgres.rs (hunk 12 above)
Comment (0xThemis) [reply oprf-key-gen/src/services/secret_manager/postgres.rs]: Yeah should not be debug print . I'd assume that the display from sqlx to respect that?

File: oprf-key-gen/src/services/key_event_watcher.rs (hunk 13 above)
Comment (0xThemis) [reply oprf-key-gen/src/services/key_event_watcher.rs:209]: Doesn't really matter now as we go to the DB on every transaction anyways which was not the case earlier. Also with retry logic this shouldn't be a problem + we will move from the deep sleep anyways according to RH

File: oprf-key-gen/src/services/secret_manager/postgres.rs (hunk 14 above)
Comment (0xThemis) [reply oprf-key-gen/src/services/secret_manager/postgres.rs:446]: I wanted to delete all because old intermediates are useless anyways and otherwise there is no logic to clean this up

File: oprf-key-gen/src/services/secret_manager/postgres.rs (hunk 15 above)
Comment (0xThemis) [reply oprf-key-gen/src/services/secret_manager/postgres.rs:146]: Yes because we set the isolation level per connection and we set it to serialization for the other one and this sticks I think. So if we acquire a connection that had previoslzy set to serial we would get it here as well.

Comment (0xThemis) [reply oprf-key-gen/src/services/secret_manager/postgres.rs:146]: Also being explicit doesnt hurt imo

File: oprf-key-gen/src/services/secret_manager/postgres.rs (hunk 18 above)
Comment (0xThemis) [reply oprf-key-gen/src/services/secret_manager/postgres.rs:237]: See reply above regarding deletes

**Review by dkales — APPROVED** [review]
